tools/stress-cluster/services/Stress.Watcher/src/JobEventHandler.cs

```diff
@@ -0,0 +1,237 @@
+                .CreateLogger();
+        }
+
+        public async Task Watch(CancellationToken cancellationToken)
+        {
+            string resourceVersion = null;
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                try
+                {
+                    var listTask = Client.ListJobForAllNamespacesWithHttpMessagesAsync(
+                        allowWatchBookmarks: true,
+                        watch: true,
+                        resourceVersion: resourceVersion,
+                        cancellationToken: cancellationToken
+                    );
+                    var tcs = new TaskCompletionSource();
+                    using var watcher = listTask.Watch<V1Job, V1JobList>(
+                        (type, pod) => {
+                            resourceVersion = pod.ResourceVersion();
+                            HandleJobEvent(type, pod);
+                        },
+                        (err) =>
+                        {
+                            Logger.Error(err, "Handling error event for job watch stream.");
+                            if (err is KubernetesException kubernetesError)
+                            {
+                                // Handle "too old resource version"
+                                if (string.Equals(kubernetesError.Status.Reason, "Expired", StringComparison.Ordinal))
+                                {
+                                    resourceVersion = null;
+                                }
+                            }
+                            tcs.TrySetException(err);
+                            throw err;
+                        },
+                        () => {
+                            Logger.Warning("Job watch has closed.");
+                            tcs.TrySetResult();
+                        }
+                    );
+                    using var registration = cancellationToken.Register(watcher.Dispose);
+                    await tcs.Task;
+                }
+                catch (Exception ex)
+                {
+                    Log.Error(ex, "Error with job watch stream.");
+                    await Task.Delay(1000, cancellationToken);
+                }
+            }
+        }
+
+        public void HandleJobEvent(WatchEventType type, V1Job job)
+        {
+            using (LogContext.PushProperty("namespace", job.Namespace()))
+            using (LogContext.PushProperty("job", job.Name()))
+            {
+                DeleteResources(job).ContinueWith(t =>
+                {
+                    if (t.Exception != null)
+                    {
+                        Logger.Error(t.Exception, "Error deleting resources.");
+                    }
+                });
+            }
+        }
+
+        public async Task DeleteResources(V1Job job)
+        {
+            if (!ShouldDeleteResources(job))
+            {
+                Logger.Debug($"Skipping resource deletion.");
```

Comment:
Should we move this logger.debug in the ShouldDeleteResources ?

Reply:
If we put it inside the method, then we would have to log it in multiple places since there are multiple return points in the function. There's an argument for just removing it, and logging at each conditional return in `ShouldDeleteResources` where the log is specific about why we're returning false.

Reply:
Yeah, that's what I meant is to have those reasons in place. so less duplicated logs, because you do have some logs in certain situations in ShouldDelete, and more clarity